### ISSUE TITLE
Make rust first class extension language.

### ISSUE THREAD

**Issue by yyoncho** [emacs-ng/emacs-ng#77]
ATM you can use rust only as a core language. It will be nice if you can develop rust packages and load them from emacs. This is similar to @ubolonton's Rust Modules but without the dynamic modules overhead and with richer access to emacs internals.

**Comment by ubolonton** [comment]
Do you mean making it easier to write crates that can be bundled in the Emacs build (static linking), or improving support for dynamically-loaded libs (dynamic modules)?

> without the dynamic modules overhead

I think a big part of the overhead is around handling non-local exits (`signal` and `throw`). Emacs uses setjmp+longjmp for these, so most of the APIs in `emacs-module.h` have to catch them to prevent jumping over foreign stack frames (in this case Rust). In the other direction, `emacs-module-rs` has to use `catch_unwind` to prevent panics from unwinding into non-Rust stack frames (which is UB). @brotzeit Do you know how remacs handle this?

> richer access to emacs internals.

Exposing safe Rust APIs to the internals is pretty hard, due to the above, and other issues (interpreter lock, GC). Exposing unsafe APIs is a lot more doable. It may involve cleaning up our crates so they can be used outside of the emacs-ng's source tree. Even easier would be exporting all symbols (`-rdynamic`) so that dynamic modules can use `dlsym` to get at the internal APIs.

**Comment by DavidDeSimone** [comment]
>emacs-module-rs has to use catch_unwind to prevent panics from unwinding into non-Rust stack frames (which is UB). @brotzeit > Do you know how remacs handles this?

IIUC, this is handled by this line in our Cargo.toml
```
# Allow unwinding in functions called from C. See
# https://www.reddit.com/r/rust/comments/565q1f/wrote_a_blogpost_from_my_experiences_the_arduous/d8h053m/
[profile.dev]
panic = "abort"
```

It's also applied to profile.release

**Comment by brotzeit** [comment]
Interesting, I didn't know.

**Comment by DavidDeSimone** [comment]
>Even easier would be exporting all symbols (-rdynamic) so that dynamic modules can use dlsym to get at the internal APIs.

Initially I was a big fan of this idea, and I still think it has a lot of upside, however I have some concerns with versioning. By making the entire internal API visible, that means that versioning and merging upstream becomes an issue. If I change the an implementation detail of an internal function, I may be breaking random modules in the wild.

**Comment by ubolonton** [comment]
> Initially I was a big fan of this idea, and I still think it has a lot of upside, however I have some concerns with versioning. By making the entire internal API visible, that means that versioning and merging in upstream becomes an issue. If I change the an implementation detail of an internal function, I may be breaking random modules in the wild.

Yeah, I don't think it should be the "official" way to expose the internals. I think of it as an "backdoor" that allows interested module writers to experiment. When someone comes up with an idea, they can use these APIs to write a proof-of-concept, which can then be used as the base for further discussions on how to safely expose them. We provide the tools (maybe in special "debug builds"), but make it very clear that they are unstable, and meant to be used for experimentation only. That would avoid the chicken-and-egg situations like in emacs-devel.

**Comment by ubolonton** [comment]
> IIUC, this is handled by this line in our Cargo.toml
> 
> ```
> # Allow unwinding in functions called from C. See
> # https://www.reddit.com/r/rust/comments/565q1f/wrote_a_blogpost_from_my_experiences_the_arduous/d8h053m/
> [profile.dev]
> panic = "abort"
> ```

Yeah, this solves the Rust -> C direction (just crash Emacs instead of unwinding). I'll check the C -> Rust direction.

**Comment by DavidDeSimone** [comment]
> > IIUC, this is handled by this line in our Cargo.toml
> > ```
> > # Allow unwinding in functions called from C. See
> > # https://www.reddit.com/r/rust/comments/565q1f/wrote_a_blogpost_from_my_experiences_the_arduous/d8h053m/
> > [profile.dev]
> > panic = "abort"
> > ```
> 
> Yeah, this solves the Rust -> C direction (just crash Emacs instead of unwinding). I'll check the C -> Rust direction.

For anything in javascript, I catch the errors via `internal_catch_all` at the rust layer. I am actually unaware if there was a generic solution used across remacs for other cases.

**Comment by petr-tik** [comment]
Judging from [this PR](https://github.com/emacs-ng/emacs-ng/commit/bd874fd1cb860d550235b047a0ef4116e4ea4f87), it's possible to use rust to provide lisp (but not interactive?) functions like git-init. 

2 things. Running `C-h f git-init` shows the following

```
git-init is a built-in function in ‘C source code’.
```

1. Is it possible/useful to decorate the lisp function with in `Rust source code` and somehow link the lisp def to native. 
2. Does this mean, one can integrate fd-find or ripgrep?

**Comment by DavidDeSimone** [comment]
The `#[lisp_fn]` macro can define an interactive function via the syntax `#[lisp_fn(intspec = "p")]` 

Point 1: That is something we plan on adding tracked by https://github.com/emacs-ng/emacs-ng/issues/86
Point 2: Our current rust infrastructure makes it pretty easy to integrate any existing rust functionality we can link via cargo. libgit2-rs is something we are working on now, but the project is open to integrating other libraries if they can benefit emacs.

**Comment by petr-tik** [comment]
>  Point 2: Our current rust infrastructure makes it pretty easy to integrate any existing rust functionality we can link via cargo. libgit2-rs is something we are working on now, but the project is open to integrating other libraries if they can benefit emacs.

As I outlined in my other comment, my main point atm is lack of async functionality in my existing emacs environment (mostly magit, ivy/counsel searchers like fd, rg - all over tramp), I still don't quite understand the async story of emacs-ng and where and how exactly deno can speed things up, but if there is any way I could make my current calls to magit-status or rg/fd async and prevent emacs-blocking, I would happily contribute.

**Comment by DavidDeSimone** [comment]
I'll go more in depth in the documentation tracked by https://github.com/emacs-ng/emacs-ng/issues/94. Deno has documentation on some their already built async functionality found at https://deno.land/std@0.83.0. Pretty much all of that should work out of the box, though you will need to import it. https://github.com/DavidDeSimone/ng-fuzzy-search shows the basic idea of spinning up a webworker and passing messages back and forth between the mainthread and a worker thread. As for adding native functionality to JS, that's a little more complex then I want to outline in this comment, but I'm happy to discuss it further if you want to get to hacking prior to me finishing the documentation. 

The main contributors are all active on our [gitter](https://gitter.im/emacsng/community), and I'm personally always available to answer questions if you would like to message me. This goes for any potential future contributors reading this.
